### DISCUSSION TITLE
simulate_fmu is much slower than openmodelica

### DISCUSSION THREAD

**Discussion by mekapl** [OpenModelica/OpenModelica#12904] [General]
I try to run a hydraulic model with openmodelica, it takes about 14-20 sec to finish the simulation. when I export the model to fmu and try to use simulate_fmu (from fmpy) the simulation is never finished. I tried the options of co-simulation and model exchange but it doesnt work
[LoopAutoDesigner.zip](https://github.com/user-attachments/files/17076629/LoopAutoDesigner.zip)

**Comment by casella** [comment]
Which version of OMC are you using to generate the FMU? Have you checked that the simulation setup in fmpy are the same as in OMEdit?

**Reply by mekapl** [reply]
I use the version 1.23.1 of OMEdit to generate my FMU
can you tell me which setup parameters I should check please?
thank you

**Reply by mekapl** [reply]
I use this line for simulate fmu:
result = simulate_fmu(fmu, stop_time=1.0, step_size=0.1, debug_logging=True, relative_tolerance=1e-6, solver='euler')
the fmu file is generated in Co-simulation mode

**Reply by casella** [reply]
OpenModelica natively uses a stiff, variable-step-size solver (dassl) to run simulations. Co-simulation uses by default the much cruder explicit Euler, which is bound to fail if the model is stiff and the step size is not small enough.

For your application, do you want to use the model for co-simulation, or to use it in model-exchange mode?

**Reply by casella** [reply]
BTW, hydraulic models are notorious for being horribly stiff, unless you take active counter-measures.

**Reply by mekapl** [reply]
Thank you for your reply,
I dont care if it is co-simulation of model-exchange, all I want is to run the fmu model in python. So I understood that it will be better to run it in model-exchange mode?

**Reply by casella** [reply]
Absolutely. Much more robust and efficient.

**Reply by casella** [reply]
I guess you should have some basic understanding of how these things work if you use them, though... 😅

**Reply by mekapl** [reply]
So, I try to understand the basics. I think that I used a bad word "I dont care". In fact, I meant that I can use both modes, I am not limited by one. I saw that the solvers available in simulate_fmu are CVode and Euler. I tried both but it still slow. I chose a simulation time of 1s with step size of 0.1s. I dont know if this step size is "big" for the simulation. Please how can I know if simulate_fmu successed to finish the initialization or not?

**Reply by casella** [reply]
> So, I try to understand the basics. I think that I used a bad word "I dont care".

😃 

> In fact, I meant that I can use both modes, I am not limited by one. I saw that the solvers available in simulate_fmu are CVode and Euler. I tried both but it still slow. I chose a simulation time of 1s with step size of 0.1s. I dont know if this step size is "big" for the simulation. Please how can I know if simulate_fmu successed to finish the initialization or not?

If the OpenModelica standard simulation works fine, I would first try to reproduce that result. So, use FMI-CS, and telly FMPy to use an implicit variable-step-size solver (e.g. CVODE, not sure if they have DASSL or IDA, which is a more modern version of DASSL), same tolerance as in the simulation, and set the step size as the Interval lenght of the simulation. With a variable step-size solver this is just the communication interval, the solver step size is variable, then the solution is interpolated and resampled with this interval. This way, you should succeed.

Then, you can try a whole range of combinations of solvers and FMI modes, but this would be a quite long story...

**Reply by mekapl** [reply]
So finally I found that the problem is with the pressure in the pipes. I shall add many "pressurizers" to avoid the error.
Everything works correctly now,
thank your for your help
